elastic/kibana
# [Watcher] Fix page crash when showing request of a watch with metadata

Fixes https://github.com/elastic/kibana/issues/170665

## Summary

This PR fixes the blank page when "Show request" is clicked for a watch that has a `metadata` property set. The reason for the page crash is that previously we used the `set` lodash function to update the `metadata` object which directly modifies the `metadata` object, which is part of the `payload` prop of the [RequestFlyout](https://github.com/elastic/kibana/blob/1b8236375a1d969e651eb172916e8fbb19dbeb6b/x-pack/plugins/watcher/public/application/sections/watch_edit_page/components/request_flyout.tsx#L29) component. Since the prop is changed, the RequestFlyout component is re-rendered and the `serializeJsonWatch` function is called again, and so an infinite loop is created. With the new approach, we make sure that the original `metadata` object is not modified so that we avoid infinite rerendering.


https://github.com/elastic/kibana/assets/59341489/ff76fdc7-8694-4414-a253-206e32954fee



**How to test:**
1. Create the following watch:

```
PUT _watcher/watch/watchywatch
{
  "trigger": {
    "schedule": {
      "interval": "30m"
    }
  },
  "input": {
    "search": {
      "request": {
        "search_type": "query_then_fetch",
        "indices": [
          "metricbeat-*"
        ],
        "rest_total_hits_as_int": true,
        "body": {
          "size": 0,
          "query": {
            "match_all": {}
          }
        }
      }
    }
  },
  "condition": {
    "compare": {
      "ctx.payload.hits.total": {
        "gte": 10
      }
    }
  },
  "actions": {
    "my-logging-action": {
      "logging": {
        "level": "info",
        "text": "There are {{ctx.payload.hits.total}} documents in your index. Threshold is 10."
      }
    }
  },
  "metadata": {
    "watcher_id": "watcher_test_sg",
    "description": "Watcher Test Description"
  }
}
```

2. Navigate to Watcher UI --> edit watch --> Click "Show request"
3. Verify that the request is displayed and the page doesn't crash.

Note about testing: I haven't added additional tests as there are already [unit tests](https://github.com/elastic/kibana/blob/1b8236375a1d969e651eb172916e8fbb19dbeb6b/x-pack/plugins/watcher/common/lib/serialization/serialize_json_watch.test.js#L34) for the `serializeJsonWatch` function that cover the case when there is a `metadata` property.
